dht20/dht20.go

```diff
@@ -0,0 +1,144 @@
+func New(bus drivers.I2C) Device {
+	return Device{
+		bus:     bus,
+		Address: defaultAddress, // Using the address defined in registers.go
+	}
+}
+
+// Configure sets up the device for communication and initializes the registers if needed.
+func (d *Device) Configure() error {
+	// Get the status word
+	d.data[0] = 0x71
+	err := d.bus.Tx(d.Address, d.data[:1], d.data[:1])
+	if err != nil {
+		return err
+	}
+
+	if d.data[0] != 0x18 {
+		// Initialize registers
+		err := d.initRegisters()
+		if err != nil {
+			return err
+		}
+	}
+	// Set the previous access time to the current time
+	d.prevAccessTime = time.Now()
+	return nil
+}
+
+// initRegisters initializes the registers 0x1B, 0x1C, and 0x1E to 0x00.
+func (d *Device) initRegisters() error {
+	// Initialize register 0x1B
+	d.data[0] = 0x1B
+	d.data[1] = 0x00
+	err := d.bus.Tx(d.Address, d.data[:2], nil)
+	if err != nil {
+		return err
+	}
+
+	// Initialize register 0x1C
+	d.data[0] = 0x1C
+	d.data[1] = 0x00
+	err = d.bus.Tx(d.Address, d.data[:2], nil)
+	if err != nil {
+		return err
+	}
+
+	// Initialize register 0x1E
+	d.data[0] = 0x1E
+	d.data[1] = 0x00
+	err = d.bus.Tx(d.Address, d.data[:2], nil)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Update reads data from the sensor and updates the temperature and humidity values.
+// Note that the values obtained by this function are from the previous call to Update.
+// If you want to use the most recent values, shorten the interval at which Update is called.
+func (d *Device) Update(which drivers.Measurement) error {
+	if which&drivers.Temperature == 0 && which&drivers.Humidity == 0 {
+		return nil
+	}
+
+	// Check if 80ms have passed since the last access
+	if time.Since(d.prevAccessTime) < 80*time.Millisecond {
+		return errUpdateCalledTooSoon
+	}
+
+	// Check the status word Bit[7]
+	d.data[0] = 0x71
+	err := d.bus.Tx(d.Address, d.data[:1], d.data[:1])
+	if err != nil {
+		return err
+	}
+	if (d.data[0] & 0x80) == 0 {
+		// Read 7 bytes of data from the sensor
+		err := d.bus.Tx(d.Address, nil, d.data[:7])
+		if err != nil {
+			return err
+		}
+		rawHumidity := uint32(d.data[1])<<12 | uint32(d.data[2])<<4 | uint32(d.data[3])>>4
+		rawTemperature := uint32(d.data[3]&0x0F)<<16 | uint32(d.data[4])<<8 | uint32(d.data[5])
+
+		// Convert raw values to human-readable values
+		d.humidity = float32(rawHumidity) / 1048576.0 * 100
+		d.temperature = float32(rawTemperature)/1048576.0*200 - 50
+
+		// Trigger the next measurement
+		d.data[0] = 0xAC
+		d.data[1] = 0x33
+		d.data[2] = 0x00
+		err = d.bus.Tx(d.Address, d.data[:3], nil)
+		if err != nil {
+			return err
+		}
+
+		// Update the previous access time to the current time
+		d.prevAccessTime = time.Now()
+	}
+	return nil
+}
+
+// Temperature returns the last measured temperature.
+func (d *Device) Temperature() float32 {
+	return d.temperature
+}
+
+// Humidity returns the last measured humidity.
+func (d *Device) Humidity() float32 {
+	return d.humidity
+}
```

Comment:
Ah! I'm just noticing this now- you are using floats! This would present a problem on certain devices with no FPU, causing lots of CPU cycles to be spent on processing the data. 

It was not explicit in the [Sensor PR](https://github.com/tinygo-org/drivers/pull/345), but I'd expect these methods to return an integer lest the sensor send back a binary floating point representation.

Reply:
Using integer arithmetic and representing the humidity and temperature as fixed point integers the code could take the following form (have not tested, did some basic arithmetic to get the numbers):

```go
rawHumidity := uint32(d.data[1])<<12 | uint32(d.data[2])<<4 | uint32(d.data[3])>>4
rawTemperature := uint32(d.data[3]&0x0F)<<16 | uint32(d.data[4])<<8 | uint32(d.data[5])

// humidity contains millipercent
d.humidity = int32(rawHumidity / 10) 
// temperature contains millicelsius
d.temperature = int32(rawTemperature/5) - 50000
```

If users need floating point representations (which they do) we could create a `sensor` package that contains something of this sort of logic:

```go
type Thermometer interface {
    drivers.Sensor
    // Temperature returns fixed point representation of temperature in milicelsius [mC]. 
    Temperature() int32
}

func ReadKelvin32(t Thermometer) (float32, error) {
     err := t.Update(drivers.Temperature)
     if err != nil {
        return err
     }
     v := t.Temperature()
     return float32(v)/1000 +273.15
}

// more advanced ADC configurations...
type sensorConversion struct {
    s drivers.Sensor
    gets []func() int32
    conversions []func(v int32) float32
}
```

Reply:
See https://github.com/tinygo-org/drivers/pull/332 for a previous attempt